questionpy_sdk/watcher.py

```diff
@@ -0,0 +1,196 @@
+        self._watch_path = watch_path
+
+        self._event_debouncer = EventDebouncer(_DEBOUNCE_INTERVAL, self._on_file_changes)
+
+    def start(self) -> None:
+        self._event_debouncer.start()
+
+    def stop(self) -> None:
+        if self._event_debouncer.is_alive():
+            self._event_debouncer.stop()
+            self._event_debouncer.join()
+
+    def dispatch(self, event: FileSystemEvent) -> None:
+        # filter events and debounce
+        if not self._ignore_event(event):
+            self._event_debouncer.handle_event(event)
+
+    def _on_file_changes(self, events: list[FileSystemEvent]) -> None:
+        # skip synchronization hassle by delegating this to the event loop in the main thread
+        asyncio.run_coroutine_threadsafe(self._notify_callback(), self._loop)
+
+    def _ignore_event(self, event: FileSystemEvent) -> bool:
+        """Ignores events that should not trigger a rebuild.
+
+        Args:
+            event: The event to check.
+
+        Returns:
+            `True` if event should be ignored, otherwise `False`.
+        """
+        if isinstance(event, FileOpenedEvent | FileClosedEvent):
+            return True
+
+        # ignore events events in `dist` dir
+        relevant_path = event.dest_path if isinstance(event, FileSystemMovedEvent) else event.src_path
+        try:
+            return Path(relevant_path).relative_to(self._watch_path).parts[0] == DIST_DIR
+        except IndexError:
+            return False
+
+
+class Watcher(AbstractAsyncContextManager):
+    """Watch a package source path and rebuild package/restart server on file changes."""
+
+    def __init__(self, source_path: Path, pkg_location: DirPackageLocation, host: str, port: int) -> None:
+        self._source_path = source_path
+        self._pkg_location = pkg_location
+        self._host = host
+        self._port = port
+
+        self._loop: asyncio.AbstractEventLoop
+        self._event_handler: _EventHandler | None
+        self._observer: BaseObserver | None
+        self._webserver: WebServer | None = None
+        self._on_change_condition = asyncio.Condition()
+
+    async def __aenter__(self) -> Self:
+        self._loop = asyncio.get_running_loop()
+
+        self.start()
+        return self
+
+    async def __aexit__(
+        self, exc_type: type[BaseException] | None, exc_value: BaseException | None, traceback: TracebackType | None
+    ) -> None:
+        self.stop_watching()
+
+    def start(self) -> None:
+        self.start_watching()
+        log.info("Watching '%s' for changes...", self._source_path)
+
+    def start_watching(self) -> None:
+        log.debug("Starting file watching...")
+
+        self._event_handler = _EventHandler(self._loop, self._notify, self._source_path)
+        self._observer = Observer()
+        self._event_handler.start()
+        self._observer.schedule(self._event_handler, self._source_path, recursive=True)
+        self._observer.start()
```

Comment:
Etwas schwer von außen zu erkennen, was hier der konzeptuelle Unterschied ist. Vielleicht `start` in `__aenter__` inlinen und `start_watching` und `stop_watching` privat machen?

Reply:
Die Idee war, dass man `Watcher` prinzipiell auch ohne context manager verwenden kann, bspw. wenn man selber scripted und den `Watcher` aus dem SDK verwendet. Man könnte dann etwas feingranularer steuern, was passiert.

Reply:
> Der Use-Case war halt, dass die Klasse grundsätzlich auch separat verwendet werden kann. Aber keine Ahnung, wie wichtig das ist. Hab's mal rausgenommen.

Eine allgeimeine Watcher-Klasse, die nicht an den `WebServer` gekoppelt ist, klingt auch sinnvoll, das war hier ja aber auch nicht wirklich gegeben.